File: 0046-permutations/0046-permutations.cpp
class Solution {
    vector<vector<int>> ans;
public:
    void getPermut(vector<int> &nums, int idx){
        if(idx == nums.size()){
            ans.push_back(nums);
            return;
        }
        
        for(int i=idx; i<nums.size(); i++){
            swap(nums[idx], nums[i]);
            getPermut(nums, idx+1);
            swap(nums[idx], nums[i]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        if(nums.size() == 1) return {nums};
        
        getPermut(nums, 0);
        
        return ans;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        
        def backtracking(count, path, used):
            if count == len(nums):
                res.append(path)
                return
            
            for i, num in enumerate(nums):
                if used[i]:
                    continue
                
                used[i] = True
                backtracking(count+1, path + [num], used)
                used[i] = False
                
        
        used = [False] * len(nums)
        backtracking(0, [], used)
        
        return res